spark/src/main/scala/org/apache/spark/sql/delta/UpdateExpressionsSupport.scala

```diff
@@ -114,6 +131,27 @@ trait UpdateExpressionsSupport extends CastSupport with SQLConfHelper with Analy
                   ArrayType(toEt, containsNull = true)
                 )
             }
+          case NestedMapConversion(from: MapType, to: MapType) if !Cast.canCast(from, to) =>
+            // Manually convert map values if the types are not compatible to allow schema
+            // evolution. This is slower than direct cast so we only do it when required.
+            if (!from.keyType.sameType(to.keyType)) {
+              throw DeltaErrors.unsupportedMapKeyTypeChange(
+                mapType = from,
+                fromKeyType = from.keyType,
+                toKeyType = to.keyType
+              )
+            }
+            val valueConverter: (Expression, Expression) => Expression = (key, _) =>
+              castIfNeeded(GetMapValue(fromExpression, key), to.valueType, allowStructEvolution)
+            val transformLambdaFunc = {
+              val keyVar = NamedLambdaVariable("keyVar", from.keyType, nullable = false)
+              val valueVar = NamedLambdaVariable("valueVar", from.valueType, from.valueContainsNull)
+              LambdaFunction(valueConverter(keyVar, valueVar), Seq(keyVar, valueVar))
+            }
+            cast(
+              TransformValues(fromExpression, transformLambdaFunc),
+              to.copy(valueContainsNull = true)
```

Comment:
Curious, why would the value always contain null?

Reply:
I initially replicated this from the code paths for arrays but that doesn't seem to be required. I removed it.